README
$Id: README,v 1.22 2005/01/18 18:02:16 autarch Exp $

This is Net::SSH::Perl. It contains implementations of
both the SSH1 and SSH2 protocols.

PREREQUISITES

Protocol 1

    * Math::GMP (1.04 or greater)
    * String::CRC32 (1.2 or greater)
    * Digest::MD5
    * IO::Socket

Protocol 2

    * Crypt::DSA (0.03 or greater)
    * Crypt::DH (0.01 or greater)
    * Math::Pari (2.001804 or greater)
    * MIME::Base64
    * Digest::MD5
    * Digest::SHA1
    * Digest::HMAC_MD5
    * Digest::HMAC_SHA1
    * Convert::PEM (0.05 or greater)
    * IO::Socket

It also optionally requires Digest::BubbleBabble for
generating bubble babble fingerprints, and Crypt::RSA
(1.37 or greater) if you want to use RSA key files in SSH2.

Note that RSA key files in SSH1 do not require Crypt::RSA.

INSTALLATION

Net::SSH::Perl installation is relatively straightforward. If
your CPAN shell is set up, you should just be able to do

    % perl -MCPAN -e 'install Net::SSH::Perl'

If you don't like that, you can download the distribution; the
latest version on CPAN can be found at

	http://search.cpan.org/dist/Net-SSH-Perl/

Download it, unpack it, then build it as per the usual:

    % perl Makefile.PL
    % make && make test

Then install it:

    % make install

If you have problems, read on.

The only slightly complicated bit in the installation is
that you'll need to install Crypt:: modules depending on
which ciphers you wish to use. Much work has gone into
making this simple; you should be asked a question about
which ciphers you'll use, and the modules should be
downloaded and installed automatically.

If it doesn't work, or if you don't like this option, you'll
need to do the installations manually. In which case you'll
need to install the prerequisites listed above for your
chosen protocol, plus any additional Crypt:: modules you
wish to use.

SAMPLES/TUTORIALS

Take a look at the scripts in eg/ for help and examples of
using Net::SSH::Perl.

* eg/cmd.pl is just a simple example of some of the
functionality in the module;

# eg/pssh is an ssh-like client for running commands on
remote servers and connecting to a shell on remote servers;

* eg/pscp is a very simple scp-like script;

* eg/pssh-keygen is, as you might guess, an implementation of
ssh-keygen in Perl, using the Net::SSH::Perl key libraries.
It contains almost all of the functionality in the OpenSSH
version of ssh-keygen.

* eg/remoteinteract.pl is an example of using register_handler
to interact with a remote (interactive) command.

Both pssh and pscp support a subset of the command line options
that the actual tools support; obviously, only those options
supported by Net::SSH::Perl are supported by the two scripts.

David Robins / dbrobins@cpan.org


Changes
Revision history for Net::SSH::Perl

1.34 2009.??.??

    - Rekey properly after 1 GB of data (rt.cpan.org #25044). Patch by Peter Oliver.
    - Don't try to process nonexistent or empty auth file (rt.cpan.org #41877).
    - Fix typo in croak message (rt.cpan.org #42056), thanks to jamie at audible.transient.net.
    - Move 'use base' call after Crypt module loading, per suggestion (rt.cpan.org #42051).
    - Only apply stdin if defined in SSH1 - John Payne (rt.cpan.org #42583)

1.33 2008.10.21

    - Fix open() calls (rt.cpan.org #40020)
    - Fix non-shell problem (rt.cpan.org #39980)
    - Allow full agent forwarding (rt.cpan.org #32190)
    - Handle hashed known_hosts files (Greg Sabino Mullane, rt.cpan.org #25175)

1.32 2008.10.16

    - Add IO::Handle to Perl.pm (rt.cpan.org #40057, #35985)
    - Minor test cleanups.

1.31 2008.10.02

    - New co-maintainer, Greg Sabino Mullane (TURNSTEP).
    - Prevent t/03-packet.t from hanging due to high file descriptor.
      (altblue at n0i.net, rt.cpan.org #6101)
    - Skip some tests if Math::GMP not installed (e.g. from choosing only 
      protocol 2 in Makefile.PL) (Greg Sabino Mullane, reported in 
      rt.cpan.org #25152)
    - If ENV{HOME} is not set, use getpwuid. If both fail and the dir 
      is needed, we croak. (Greg Sabino Mullane, expanded from patch 
      by dgehl at inverse.ca in rt.cpan.org #25174)
    - Fix incorrect logical/bitwise AND mixup (Peter.Haydon at uk.fujitsu.com, 
      rt.cpan.org #31490)
    - Allow empty stdin for SSH2 (rcp at rcable.co.uk, rt.cpan.org #32730)
    - Adjust terminal dimensions dynamically if Term::ReadKey is available
      (john at sackheads.org, rt.cpan.org #34874)

1.30  2005.10.12
    - Fix for local *READ/*WRITE tie problem in open2 function (Bas van
      Sisseren).

1.29  2005.10.08
    - Require Digest::SHA1 2.10 or greater (fixed problem with BSD connection)
      (Adam Callis).
    - Add suppress_shell SSH2 option for SFTP (Sean McMillan).
    - Use effective (not real) user-id to detect root (David Wheeler,
      rt.cpan.org #13373).
    - Make "ChallengeResponseAuthentication" configuration item boolean rather
      than string (mark AT blackmans.org, rt.cpan.org #14549).

1.28  2005.03.17
    - Patch to skip warning messages emitted by F-Secure sshd before the SSH
      signature (Neal Ley).
    - Fix for "Can't locate object method "blocking" via package "IO::Handle""
      error (rt.cpan.org #11674, Daniel C. Silverstein (dans AT pch.net)).

1.27  2005.02.06
    - Set socket to non-blocking in more portable manner (rt.cpan.org #6728).
    - Default port to 22 in case service list is not available (rt.cpan.org
      #5924, Gavin Brock).
    - Allow Net::SSH::Ciphers::supported to take a protocol option to get
      SSH2 ciphers (rt.cpan.org #4648, dan at midstream.com), and also change
      Net::SSH::Perl::Kex to map SSH1 names to SSH2 names.
    - Support BindAddress (bind to specific IP) (rt.cpan.org #6725, patch from
      matt at katipo.co.nz).
    - Have Makefile.PL only generate the prerequisites list, and not try to
      install any (let the install shell handle it); this allows compatibility
      with CPANPLUS and sudo installs (rt.cpan.org #6835, Michael Schwern).

1.26  2005.01.18
    - New co-maintainer, David Robins (DBROBINS).
    - Fix Math::Pari large integer leak in Net::SSH::Perl::Util::SSH2MP bin2mp
      function (has been brought to the attention of the Math::Pari maintainer
      since it's probably a Math::Pari problem, but fixed here because it's a
      big win for a small change and the same "fix" is already in mp2bin).
    - Require perl 5.6 in Makefile.PL (was '5.0006' change to '5.006').

1.25  2004.01.19
    - Fix a bizarre test failure in 03-packet.t if String::CRC32 isn't
      installed (the tests were testing SSH1 but we can't assume the
      user selected this protocol).  Reported by Damien Clark.
      rt.cpan.org ticket #4717.

1.24  2003.11.14
    - New (temporary) maintainer, Dave Rolsky.
    - Fixed a number of circular references in the SSH2
      implementation, which caused connections to be held open
      indefinitely in a persistent environment like mod_perl.  This
      uses weak references, so Perl 5.6.0+ is now required.  This work
      was funded by Kineticode, Inc.
    - Added a LICENSE file.

1.23  2001.10.04
    - Added --sdk option to Makefile.PL.
    - Added Net::SSH::Perl::SSH2::select_class so that Net::SSH::W32Perl
      can override it and reuse client_loop.

1.22  2001.08.28
    - Fixed bug with channel window sizes: Net::SSH::Perl in SSH-2 mode
      was not respecting the server's channel window size, and was
      sending data indiscriminately. It now buffers data to be sent and
      sends it when the sshd has enough window space available. Thanks
      to Megan Dorrah (and others) for the spot.
    - Fix receiving STDERR in SSH-2 protocol. This was broken. Thanks to
      Ulrich Scheid for the spot (and to Sagar Shah and Vicente J. Ruiz
      Jurado for subsequent spots :).
    - Added Channel::drain_outgoing, necessitated by the above fix to
      window sizes; that fix meant that output buffers were not being
      drained before EOF was sent on the channel. drain_outgoing must
      now be called before EOF is sent; it continually sends outgoing
      data, and checks for increased window sizes, until there is no more
      data to send. This will be done automatically, of course.
    - Display SSH-2 BANNER messages during authentication if the session
      is interactive.
    - Net::SSH::Perl::Packet now uses the third argument to syswrite
      (LENGTH). This was necessary to make the module work in Perl
      5.005_2, apparently. Thanks to Brian Carpenter for the spot.
    - Added -2 and -1 options (SSH-2 and SSH-1, respectively) to
      eg/pscp.

1.21  2001.07.18
    - Fixed Handle::SSH(?:1|2)::WRITE when sending large amounts of data,

Makefile.PL
# $Id: Makefile.PL,v 1.29 2008/09/24 19:21:20 turnstep Exp $

use strict;
use 5.006;
use ExtUtils::MakeMaker qw(prompt WriteMakefile);

my %prereq = (
        'Digest::MD5'       => 0,
        'IO::Socket'        => 0,
);

my %SSH_PREREQ = (
    1 => {
        'String::CRC32'     => '1.2',
        'Math::GMP'         => '1.04',
        'Scalar::Util'      => 0,
    },

    2 => {
        'Digest::SHA1'      => '2.10',
        'Digest::HMAC_MD5'  => 0,
        'Digest::HMAC_SHA1' => 0,
        'Crypt::DSA'        => '0.11',
        'Crypt::DH'         => '0.01',
        'Math::Pari'        => '2.001804',
        'MIME::Base64'      => 0,
        'Convert::PEM'      => '0.05',
    },
);

$SSH_PREREQ{3} = { map %{$SSH_PREREQ{$_}}, 1..2 };

print<<MSG;
This is Net::SSH::Perl.

As of version 1.00, Net::SSH::Perl supports both the SSH1 and
SSH2 protocols natively. The two protocols have different
module prerequisitives, so you need to decide which protocol(s)
you plan to use. If you use one or the other, only those modules
for your chosen protocol will be installed; if you choose both,
all of the supporting modules will be installed. Please choose
the protocols you'd like to use from the following list ("Both"
is the default).

MSG

printf "    [%d] SSH%d\n", $_, $_ for 1..2;
printf "    [3] Both SSH1 and SSH2\n";

my $p = prompt("\nWhich protocol(s) do you plan to use?", 3);
print "\n";

@prereq{keys %{$SSH_PREREQ{$p}}} = values %{$SSH_PREREQ{$p}};

my @cryptmod = (
    [ 'IDEA', 'Crypt::IDEA' ],
    [ 'DES','Crypt::DES' ],
    [ 'DES3', 'Crypt::DES' ],
    [ 'Blowfish', 'Crypt::Blowfish' ],
    [ 'RC4', '' ],
);

print<<MSG;
Some of the Net::SSH::Perl ciphers depend on a Crypt:: module from
CPAN. You may already have the necessary modules installed, in which
case you don't need to bother with this step. Otherwise you'll need
to install at least one cipher to use Net::SSH::Perl. Please choose
at least one from the following list (Crypt::IDEA is the default).

MSG

my $i = 1;
for my $ciph(sort { $a->[0] <=> $b->[0] } @cryptmod) {
    printf "    [%d] %s\n", $i++, $ciph->[0];
}
my $c = prompt("\nEnter your choices, separated by spaces:", 1);
print "\n";

for my $id(split /\s+/, $c) {
    next unless $cryptmod[$id-1]->[1];
    $prereq{$cryptmod[$id-1]->[1]} = '0';
}

print "Checking for optional modules\n\n";

unless (have_module('Digest::BubbleBabble', 0.01)) {
    print<<MSG, "\n";
Digest::BubbleBabble is required if you want to generate bubble babble
key fingerprints with pssh-keygen.
MSG

    if (read_yes_or_no("Would you like to install it? (y/n)", "y")) {
        $prereq{'Digest::BubbleBabble'} = '0.01';
    }
    print "\n";
}

unless (have_module('Crypt::RSA', 1.37)) {
    print<<MSG, "\n";
Crypt::RSA is required if you wish to use the ssh-rsa public
key algorithm (ssh-dss is used by default).
MSG
    if (read_yes_or_no("Would you like to install it? (y/n)", "y")) {
        $prereq{'Crypt::RSA'} = '1.37';
    }
    print "\n";
}

WriteMakefile(
    NAME         => 'Net::SSH::Perl',
    DISTNAME     => 'Net-SSH-Perl',
    VERSION_FROM => 'lib/Net/SSH/Perl.pm',
    PREREQ_PM    => \%prereq,
    AUTHOR       => 'David Robins <dbrobins@cpan.org>',
    ABSTRACT     => 'Perl client interface to SSH',
    NO_META      => 1,
);

sub read_yes_or_no {
    my($prompt, $def) = @_;
    my $ans = prompt($prompt, $def);
    $ans =~ /^y/i;
}

sub have_module {
    my($name, $ver) = @_;
    eval("use $name" . ($ver ? " $ver;" : ";"));
    !$@;
}
